### PR TITLE
[klippa] closure for COLR table

### PR DESCRIPTION
**Pull request by qxliu76**
Planning for COLR is not done yet. Things left: remapping for layer/palette indices, varstore planning.

### CODE REVIEW

File: read-fonts/src/tables/colr.rs [hunk 1]
```diff
@@ -101,4 +157,686 @@ impl<'a> Colr<'a> {
...
+        let offset = (paint.offset_data().as_bytes().as_ptr() as usize - self.colr_head) as u32;
+        if self.visited_paints.contains(offset) {
+            return true;
+        }
+
+        self.visited_paints.insert(offset);
+        false
+    }
+
+    pub fn add_layer_indices(&mut self, first_layer_index: u32, last_layer_index: u32) {
```
Comment (cmyr) [inline read-fonts/src/tables/colr.rs:267]: does this need to be public API? Ditto for some of the other methods here, I think they're only called internally?

File: read-fonts/src/tables/colr.rs [hunk 2]
```diff
@@ -101,4 +157,686 @@ impl<'a> Colr<'a> {
...
+impl<'a> ColorLine<'a> {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
+        for colorstop in self.color_stops() {
+            colorstop.v1_closure(c);
+        }
+    }
+}
+
+impl<'a> VarColorLine<'a> {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
```
Comment (cmyr) [inline read-fonts/src/tables/colr.rs:313]: this is starting to look an awful lot like a trait...

**Review by cmyr — APPROVED** [review]
Looks good, a few comments inline but nothing blocking :)

File: read-fonts/src/tables/colr.rs [hunk 3]
```diff
@@ -37,6 +39,60 @@ impl<'a> Colr<'a> {
         Ok((layer.glyph_id(), layer.palette_index()))
     }
 
+    /// Collect the transitive closure of V0 glyphs needed for all of the input glyphs set
+    pub fn v0_closure_glyphs(
+        &self,
+        glyph_set: &IntSet<GlyphId>,
+        glyphset_colrv0: &mut IntSet<GlyphId>,
+    ) {
+        if let Some(Ok(records)) = self.base_glyph_records() {
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply read-fonts/src/tables/colr.rs:48]: If that change is made then glyphset_colrv0.union(glyph_set) will need to be moved up to before this line.

File: klippa/src/lib.rs [hunk 4]
```diff
@@ -34,6 +34,7 @@ const MAX_GID: GlyphId = GlyphId::new(0xFFFFFF);
 pub struct Plan {
     unicodes: IntSet<u32>,
     glyphset_gsub: IntSet<GlyphId>,
+    glyphset_colred: IntSet<GlyphId>,
```
[the remark this answers is not in the document]
Comment (qxliu76) [reply klippa/src/lib.rs:37]: the name is coming from fonttools, meaning it include glyphs after COLR closure

File: read-fonts/src/tables/colr.rs (hunk 2 above)
Comment (qxliu76) [reply read-fonts/src/tables/colr.rs:314]: yes, I assume this is fine cause we don't need to use trait bound in functions

File: read-fonts/src/tables/colr.rs [hunk 5]
```diff
@@ -37,6 +39,61 @@ impl<'a> Colr<'a> {
...
+    }
+
+    //Collect the transitive closure of V0 palette indices needed for all of the input glyphs set
+    //It's similar to closure glyphs but in a separate fn, because v1 closure might adds more v0 glyphs, so this fn needs to be called after v1 closure
+    pub fn v0_closure_palette_indices(
+        &self,
+        glyph_set: &IntSet<GlyphId>,
+        palette_indices: &mut IntSet<u16>,
+    ) {
+        if let Some(Ok(records)) = self.base_glyph_records() {
```
Comment (garretrieger) [inline read-fonts/src/tables/colr.rs:77]: This could be a let ... else { return }; to reduce nesting.

File: read-fonts/src/tables/colr.rs [hunk 6]
```diff
@@ -101,4 +158,686 @@ impl<'a> Colr<'a> {
...
+}
+
+impl<'a> PaintColrLayers<'a> {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
+        let num_layers = self.num_layers();
+        if num_layers == 0 {
+            return;
+        }
+
+        if let Some(Ok(layer_list)) = c.colr.layer_list() {
```
Comment (garretrieger) [inline read-fonts/src/tables/colr.rs:367]: here to.

File: read-fonts/src/tables/colr.rs [hunk 7]
```diff
@@ -101,4 +158,686 @@ impl<'a> Colr<'a> {
...
+        if let Ok(paint) = self.paint() {
+            c.dispatch(&paint);
+        }
+    }
+}
+
+impl<'a> PaintColrGlyph<'a> {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext) {
+        let glyph_id = self.glyph_id();
+        if let Some(Ok(list)) = c.colr.base_glyph_list() {
```
Comment (garretrieger) [inline read-fonts/src/tables/colr.rs:460]: here too

File: read-fonts/src/tables/colr.rs [hunk 8]
```diff
@@ -101,4 +158,686 @@ impl<'a> Colr<'a> {
...
+
+        if let Ok(backdrop_paint) = self.backdrop_paint() {
+            c.dispatch(&backdrop_paint);
+        }
+    }
+}
+
+impl Clip {
+    fn v1_closure(&self, c: &mut Colrv1ClosureContext, clip_list: &ClipList) {
+        if let Ok(clip_box) = self.clip_box(clip_list.offset_data()) {
```
Comment (garretrieger) [inline read-fonts/src/tables/colr.rs:666]: here too

**Review by garretrieger — APPROVED** [review]

**Pull request merged**